Documentation: Add section to validate Hubble TLS is enabled (#34416)
cilium-team/cilium: Documentation/observability/hubble/configuration/tls.rst

        the cluster name defined by ``cluster.name`` (defaults to ``default``).

Validating the Installation
---------------------------

The following section guides you through validating that TLS is enabled for Hubble
and the connection between Hubble Relay and Hubble Server is using mTLS to secure the session.
Additionally, the commands below can be used to troubleshoot issues with your TLS configuration if you encounter any issues.

Start by creating a Hubble CLI pod within the namespace that Hubble components are running in (for example: ``kube-system``):

.. code-block:: shell-session

    $ kubectl apply -n kube-system -f https://raw.githubusercontent.com/cilium/cilium/main/examples/hubble/hubble-cli.yaml

List Hubble Servers by running ``hubble watch peers`` within the newly created pod:

.. code-block:: shell-session

    $ kubectl exec -it -n kube-system deployment/hubble-cli -- \
    hubble watch peers --server unix:///var/run/cilium/hubble.sock

    PEER_ADDED   172.18.0.2 kind-worker (TLS.ServerName: kind-worker.default.hubble-grpc.cilium.io)
    PEER_ADDED   172.18.0.3 kind-control-plane (TLS.ServerName: kind-control-plane.kind.hubble-grpc.cilium.io)

> Copy the IP and the server name of the first peer into the following environment variables for the next steps

Reviewer: ```suggestion
Copy the IP and the server name of the first peer into the following environment variables for the next steps:
```

.. code-block:: shell-session

    $ IP=172.18.0.2
    $ SERVERNAME=kind-worker.default.hubble-grpc.cilium.io

Connect to the first peer with the Hubble Relay client certificate to confirm

> that the Hubble server is working when presented the right certificate:

Reviewer: Wording suggestion to make this more specific:

```suggestion
that the Hubble server is accepting connections from clients who present
the correct certificate:
```

.. code-block:: shell-session

    $ kubectl exec -it -n kube-system deployment/hubble-cli -- \
    hubble observe --server ${IP?}:4244 \
        --tls-server-name ${SERVERNAME?} \
        --tls-ca-cert-files /var/lib/hubble-relay/tls/hubble-server-ca.crt \
        --tls-client-cert-file /var/lib/hubble-relay/tls/client.crt \
        --tls-client-key-file /var/lib/hubble-relay/tls/client.key

>     Dec 13 08:49:58.888: 10.20.1.124:60588 (host) -> kube-system/coredns-565d847f94-pp8zs:8181 (ID:7518) to-endpoint FORWARDED (TCP Flags: SYN)

Reviewer: Is there a missing newline here? The other `shell-session` blocks in this doc use a newline between the command and the output.
```suggestion
        --tls-client-key-file /var/lib/hubble-relay/tls/client.key

    Dec 13 08:49:58.888: 10.20.1.124:60588 (host) -> kube-system/coredns-565d847f94-pp8zs:8181 (ID:7518) to-endpoint FORWARDED (TCP Flags: SYN)
```

    ...
    ...

Now try to query the Hubble server without providing any client certificate:

.. code-block:: shell-session

    $ kubectl exec -it -n kube-system deployment/hubble-cli -- \
    hubble observe --server ${IP?}:4244 \
        --tls-server-name ${SERVERNAME?} \
        --tls-ca-cert-files /var/lib/hubble-relay/tls/hubble-server-ca.crt

    failed to connect to '172.18.0.2:4244': context deadline exceeded: connection error: desc = "transport: authentication handshake failed: mTLS client certificate requested, but not provided"
    command terminated with exit code 1

You can also try to connect without TLS:

.. code-block:: shell-session

    $ kubectl exec -it -n kube-system deployment/hubble-cli -- \
    hubble observe --server ${IP?}:4244

    failed to connect to '172.18.0.2:4244': context deadline exceeded: connection error: desc = "error reading server preface: EOF"
    command terminated with exit code 1

To troubleshoot the connection, install OpenSSL in the Hubble CLI pod:

.. code-block:: shell-session

    $ kubectl exec -it -n kube-system deployment/hubble-cli -- apk add --update openssl

> Then OpenSSL to get more details:

Reviewer: Wording suggestion to make this more specific:

```suggestion
Then, use OpenSSL to connect to the Hubble server get more details about the TLS handshake:
```

        0000 - 55 93 99 70 30 37 6a 77-43 d7 0c 34 9f 24 51 40   U..p07jwC..4.$Q@
        ...
        ...
        0690 - 11 6d 26 ec 99 3a 6e a9-56 c9 ad a0 49 e2 f5 6a   .m&..:n.V...I..j

Press ``ctrl-d`` to signal the TLS session and connection should be terminated. After the session is over you will see output similar to the following:

.. code-block:: shell-session

    @DONE
        06a0 - bf eb 8b 1d 8d 43 46 2a-07 02 e1 44 35 45 b1 a0   .....CF*...D5E..
        06b0 - 7d bb 27 2f 1a 35 b2 da-0d 00 15 fd 6c 1f 00 3b   }.'/.5......l..;
        06c0 - 9a 6e ff c9 5d ad 6b af-f7 20 39 99 5b ae 72 03   .n..].k.. 9.[.r.
        06d0 - c8 2d 93 7a e5 a7 e0 d5-70 95 8f b5 0b 56 9c      .-.z....p....V.

        Start Time: 1723744378
        Timeout   : 7200 (sec)
        Verify return code: 0 (ok)
        Extended master secret: no
        Max Early Data: 0
    ---
    read R BLOCK

The output of this OpenSSL command is similar to the previous output, but without the error message.
There is also an additional section ``Post-Handshake New Session Ticket arrived``, indicating that an TLS session was established.
This confirms that the client certificate is valid.

> Additionally when the connection ended it printed a summary of the TLS session.

Reviewer: Some wording suggestions, let me know what you think:

```suggestion
There is also an additional section, named``Post-Handshake New Session Ticket arrived``, the presence
of which indicates that the client certificate is valid and a TLS session was established.
The summary of the TLS session printed after the connection has ended can also be used as an
indicator of the established TLS session.
```

Author: This looks better.

        the cluster name defined by ``cluster.name`` (defaults to ``default``).

Validating the Installation
---------------------------

The following section guides you through validating that TLS is enabled for Hubble
and the connection between Hubble Relay and Hubble Server is using mTLS to secure the session.
Additionally, the commands below can be used to troubleshoot issues with your TLS configuration if you encounter any issues.

Start by creating a Hubble CLI pod within the namespace that Hubble components are running in (for example: ``kube-system``):

.. code-block:: shell-session

    $ kubectl apply -n kube-system -f https://raw.githubusercontent.com/cilium/cilium/main/examples/hubble/hubble-cli.yaml

List Hubble Servers by running ``hubble watch peers`` within the newly created pod:

.. code-block:: shell-session

    $ kubectl exec -it -n kube-system deployment/hubble-cli -- \
    hubble watch peers --server unix:///var/run/cilium/hubble.sock

    PEER_ADDED   172.18.0.2 kind-worker (TLS.ServerName: kind-worker.default.hubble-grpc.cilium.io)
    PEER_ADDED   172.18.0.3 kind-control-plane (TLS.ServerName: kind-control-plane.kind.hubble-grpc.cilium.io)

> Copy the IP and the server name of the first peer into the following environment variables for the next steps:

Reviewer 2: This step is assuming TLS is correctly setup, we should add a section showing the output of `hubble watch peers` when tls is disabled and what to look for (e.g. `TLS.ServerName` missing from the output). Maybe adding a step on the way to check whether hubble TLS is enabled in the cilium configMap since we are assuming access to `kube-system`?

Author: I've added a note just below this line mentioning the `TLS.ServerName` and to refer to previous steps to enable TLS if it's missing. I also added a step at the top of the troubleshooting section here to check the cilium-config as well.

.. code-block:: shell-session

    $ IP=172.18.0.2
    $ SERVERNAME=kind-worker.default.hubble-grpc.cilium.io

Connect to the first peer with the Hubble Relay client certificate to confirm

> that the Hubble server is that the Hubble server is accepting connections from

Reviewer: Small typo:

```suggestion
that the Hubble server is accepting connections from
```

        PSK identity: None
        PSK identity hint: None
        SRP username: None
        TLS session ticket lifetime hint: 604800 (seconds)
        TLS session ticket:
        0000 - 55 93 99 70 30 37 6a 77-43 d7 0c 34 9f 24 51 40   U..p07jwC..4.$Q@
        ...
        ...
        0690 - 11 6d 26 ec 99 3a 6e a9-56 c9 ad a0 49 e2 f5 6a   .m&..:n.V...I..j

Press ``ctrl-d`` to signal the TLS session and connection should be terminated.
After the session has ended you will see output similar to the following:

.. code-block:: shell-session

    @DONE
        06a0 - bf eb 8b 1d 8d 43 46 2a-07 02 e1 44 35 45 b1 a0   .....CF*...D5E..
        06b0 - 7d bb 27 2f 1a 35 b2 da-0d 00 15 fd 6c 1f 00 3b   }.'/.5......l..;
        06c0 - 9a 6e ff c9 5d ad 6b af-f7 20 39 99 5b ae 72 03   .n..].k.. 9.[.r.
        06d0 - c8 2d 93 7a e5 a7 e0 d5-70 95 8f b5 0b 56 9c      .-.z....p....V.

        Start Time: 1723744378
        Timeout   : 7200 (sec)
        Verify return code: 0 (ok)
        Extended master secret: no
        Max Early Data: 0
    ---
    read R BLOCK

The output of this OpenSSL command is similar to the previous output, but without the error message.

> There is also additional an additional section, starting with ``Post-Handshake

Reviewer: Small typo:

```suggestion
There is also an additional section, starting with ``Post-Handshake
```
